src/bin/common/solving_loop.ml

```diff
@@ -524,6 +595,20 @@ let main () =
     |> Typer_Pipe.init ~type_check
   in
 
+  (* Initializing hooks in the mode handler.
+     When we perform a check-sat, the environment we are working on is specific
+     to the Sat or Unsat mode we end up in. If we start asserting again, we must
+     do it in the previous environment.
+  *)
+  let () =
+    DO.Mode.reset_hooks ();
+    DO.Mode.add_hook
+      (fun _ ~old ~new_ st ->
+         match old, new_ with
+         | (Sat | Unsat), Assert -> internal_pop st
+         | _ -> st
```

Comment:
I am quite uncomfortable with this approach. Implicitly popping the state when we switch mode seems brittle, because we need to be very careful with the way state transitions are handled. We kind of have to anyways, but I'd very much rather have code like `prepare_for_query st` or `pop_model st` or the like to explicitly perform the check rather than an innocuous-looking `DO.Mode.set Assert st` that needs accompanying comments to clarify it is only there to trigger hooks. Just call the code.

Reply:
It is harder to make sure the `pop` is called each time we change mode than adding a hook. Besides, the exact semantic of the internal push/pop is precisely to be called whenever we switch mode. Of course we will trigger errors when we forget to change modes, but that is IMO other issues. I have tried to document the code as well as possible about this.

Another solution I have thought about is to add another mode: `Computing` between `Assert` and `Sat`/`Unsat`. When going from `Assert` to `Computing`, we perform an internal push ; when we go from `Sat`/`Unsat` to `Assert`, we perform an internal pop. That way, all calls to internal pop/push are centralized in hooks, and we can even add a flag in the state to enforce the consistency.

Reply:
> the exact semantic of the internal push/pop is precisely to be called whenever we switch mode

I disagree: there is now an additional transition to the `Assert` mode when calling `(check-sat)` which is not present in the SMT-LIB automaton, which muddies the semantics. My issue is exactly that the internal push/pop thing doesn't map properly to the existing state machine, which is the part that makes me uncomfortable (I should have commented there rather than here, sorry).

I kind of like having a separate `Computing` or `Check` state because it makes the semantics clear (if we are in `Computing` or `Check` we have the internal push, otherwise we don't — as you say, it would make the internal push/pops actually follow the state transitions), but I also don't really like introducing a new mode for this because it also makes the mode deviate from the SMT-LIB automaton it is supposed to represent.

I am aware that not calling `DO.Mode.set Assert` there would by risky and could cause double pops. As an alternative to the `Computing` state, what do you think of having a separate field `has_local_state`? We'd set it to `true` when an internal push is needed (due to the presence of local hypotheses or assertions in `Solve` — but we wouldn't for `` `Solve ([], [])`` which is `(check-sat)`); if it is true when entering `Assert` mode we do an `internal_pop` and set it to `false` (so that it is always `false` in `Assert` mode); if it is true in `handle_query`, we do an `internal_pop` and set it to `false` as well. This way, the hooks are used to enforce changes that are part of the SMT-LIB automaton, and explicit calls to `internal_push` and `internal_pop` are made otherwise.

Reply:
> I disagree: there is now an additional transition to the Assert mode when calling (check-sat) which is not present in the SMT-LIB automaton, which muddies the semantics. 

You're right, I shouldn't do this transition and have a flag that states whether I already have pushed or not; which is equivalent to the `has_local_state` field you are describing.